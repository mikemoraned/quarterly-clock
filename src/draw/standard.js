import * as d3 from "d3";

export function defaultPositionRounding(value) {
    const rounded = value.toFixed(2);
    return parseFloat(rounded);
}

export function defaultRotationRounding(value) {
    const rounded = value.toFixed(2);
    return parseFloat(rounded);
}

export function defaultFontSizeFormat(value) {
    const rounded = value.toFixed(2);
    return `${rounded}px`;
}

export function defaultArcGenerator() {
    return d3.arc().digits(3);
}