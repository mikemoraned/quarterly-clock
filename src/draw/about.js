import * as d3 from "d3";
import { defaultPositionRounding as dpr } from "./standard";

export function drawLogo(dataModel, guidesModel, svg) {
  const parentGroup = svg.selection.append("g").attr("id", "logo");

  const positionForQuarterIndex = [
    guidesModel.logo.bottom,
    guidesModel.logo.top,
    guidesModel.logo.top,
    guidesModel.logo.bottom,
  ];
  const position = positionForQuarterIndex[dataModel.currentQuarter.index];

  const path = {
    startX: position.x - guidesModel.outerRadius / 2.0,
    startY: position.y + guidesModel.outerRadius / 5.0,
    radiusX: guidesModel.outerRadius,
    radiusY: guidesModel.outerRadius,
    endX: position.x + guidesModel.outerRadius / 2.0,
    endY: position.y + guidesModel.outerRadius / 5.0,
  };
  const d = `M ${dpr(path.startX)},${dpr(path.startY)} A ${dpr(path.radiusX)},${dpr(path.radiusY)} 0 0,1 ${dpr(path.endX)},${dpr(path.endY)}`;
  parentGroup
    .append("defs")
    .append("path")
    .attr(
      "d",
      d
    )
    .attr("id", "logoPath");

  parentGroup.append("path")
    .attr(
      "d",
      d
    ).attr("class", "guide")
    .style("fill", "red")
    .style("stroke", "black");

  const text = parentGroup
    .append("a")
    .attr("href", "https://github.com/mikemoraned/quarterly-clock")
    .append("text");

  text
    .attr(
      "style",
      `font-size: ${guidesModel.logo.fontSize}; dominant-baseline: auto; text-anchor: middle`
    )
    .attr("fill", "black");

  const textPath = text
    .append("textPath")
    .attr("startOffset", "50%")
    .attr("xlink:href", "#logoPath");

  textPath
    .append("tspan")
    .text("\uf017")
    .attr("style", `font-family: "Font Awesome 5 Free"`);
  textPath.append("tspan").text("uarterly");
}

export function drawReadme(dataModel, guidesModel, svg) {
  const parentGroup = svg.selection.append("g").attr("id", "readme");

  const positionForQuarterIndex = [
    guidesModel.readme.bottom,
    guidesModel.readme.top,
    guidesModel.readme.top,
    guidesModel.readme.bottom,
  ];
  const position = positionForQuarterIndex[dataModel.currentQuarter.index];

  const path = {
    startX: position.x - guidesModel.outerRadius / 2.0,
    startY: position.y + guidesModel.outerRadius / 5.0,
    radiusX: guidesModel.outerRadius,
    radiusY: guidesModel.outerRadius,
    endX: position.x + guidesModel.outerRadius / 2.0,
    endY: position.y + guidesModel.outerRadius / 5.0,
  };
  const d = `M ${dpr(path.startX)},${dpr(path.startY)} A ${dpr(path.radiusX)},${dpr(path.radiusY)} 0 0,1 ${dpr(path.endX)},${dpr(path.endY)}`
  parentGroup
    .append("defs")
    .append("path")
    .attr(
      "d",
      d
    )
    .attr("id", "readmePath");

  parentGroup.append("path")
    .attr(
      "d",
      d
    ).attr("class", "guide")
    .style("fill", "red")
    .style("stroke", "black");

  const text = parentGroup
    .append("a")
    .attr(
      "href",
      "https://github.com/mikemoraned/quarterly-clock/blob/main/README.md"
    )
    .append("text");

  text
    .attr(
      "style",
      `font-size: ${guidesModel.readme.fontSize}; dominant-baseline: auto; text-anchor: middle`
    )
    .attr("fill", "black");

  const textPath = text
    .append("textPath")
    .attr("startOffset", "50%")
    .attr("xlink:href", "#readmePath");

  textPath
    .append("tspan")
    .text("\uf4d5")
    .attr("style", `font-family: "Font Awesome 5 Brands"`);
  textPath.append("tspan").text(" Readme");
}
