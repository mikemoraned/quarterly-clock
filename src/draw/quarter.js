import * as d3 from "d3";
import { defaultPositionRounding as dpr, defaultArcGenerator } from "./standard";

export function drawCurrentQuarter(dataModel, guidesModel, svg) {
  const arcGenerator = d3.arc();

  console.log("D3 version:", d3.version);

  arcGenerator
    .innerRadius(guidesModel.outerRadius / 6)
    .outerRadius(guidesModel.outerRadius - guidesModel.outerRadius / 4)
    .digits(2);

  const arc = {
    startAngle: dataModel.currentQuarter.start.yearFraction * 2.0 * Math.PI,
    endAngle: dataModel.currentQuarter.end.yearFraction * 2.0 * Math.PI,
  };

  const parentGroup = svg.selection.append("g").attr("id", "current-quarter");

  parentGroup
    .append("path")
    .attr("class", "current-quarter")
    .attr("d", arcGenerator(arc))
    .attr(
      "fill",
      guidesModel.colors.quarters[dataModel.currentQuarter.index].color
    )
    .attr("stroke", "none");
}

export function drawAllQuarterContext(dataModel, guidesModel, svg) {
  const arcGenerator = d3.arc();

  arcGenerator
    .innerRadius(guidesModel.outerRadius / 6)
    .outerRadius(guidesModel.outerRadius - 0.7 * guidesModel.outerRadius);

  const arcs = [
    {
      startAngle: 0,
      endAngle: 0.5 * Math.PI,
      color: guidesModel.colors.quarters[0].color,
    },
    {
      startAngle: 0.5 * Math.PI,
      endAngle: Math.PI,
      color: guidesModel.colors.quarters[1].color,
    },
    {
      startAngle: Math.PI,
      endAngle: 1.5 * Math.PI,
      color: guidesModel.colors.quarters[2].color,
    },
    {
      startAngle: 1.5 * Math.PI,
      endAngle: 2.0 * Math.PI,
      color: guidesModel.colors.quarters[3].color,
    },
  ];

  const parentGroup = svg.selection.append("g").attr("id", "all-quarters");

  arcs.forEach((arc) => {
    parentGroup
      .append("path")
      .attr("d", arcGenerator(arc))
      .attr("fill", arc.color)
      .attr("stroke", "none");
  });
}

export function drawQuarterLabel(dataModel, guidesModel, svg) {
  const parentGroup = svg.selection.append("g").attr("id", "info");

  const positionForQuarterIndex = [
    guidesModel.quarterLabel.left,
    guidesModel.quarterLabel.left,
    guidesModel.quarterLabel.right,
    guidesModel.quarterLabel.right,
  ];

  const position = positionForQuarterIndex[dataModel.currentQuarter.index];

  parentGroup
    .append("text")
    .text(`${dataModel.currentQuarter.label}`)
    .attr("x", position.x)
    .attr("y", position.y)
    .attr(
      "style",
      `font-size: ${guidesModel.quarterLabel.fontSize}; dominant-baseline: text-bottom; text-anchor: left`
    )
    .attr(
      "fill",
      guidesModel.colors.quarters[dataModel.currentQuarter.index].color
    )
    .attr("stroke", "none");
}
