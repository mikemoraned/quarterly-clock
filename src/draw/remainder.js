import * as d3 from "d3";
import { defaultArcGenerator, defaultPositionRounding as dpr } from "./standard.js";

export function drawRemainder(dataModel, guidesModel, svg) {
  const pattern = svg.root
    .append("defs")
    .append("pattern")
    .attr("id", "remainder-pattern")
    .attr("width", `${guidesModel.remainder.box.width}`)
    .attr("height", `${guidesModel.remainder.box.height}`)
    .attr("patternUnits", "userSpaceOnUse")
    .attr("patternTransform", "rotate(20)");

  pattern
    .append("rect")
    .attr("width", `${guidesModel.remainder.box.width}`)
    .attr("height", `${guidesModel.remainder.box.height}`)
    .attr("opacity", "0.8")
    .attr("fill", guidesModel.colors.remainder.bg.color);

  const extraPattern = svg.root
    .append("defs")
    .append("pattern")
    .attr("id", "remainder-pattern-extra")
    .attr("width", `${guidesModel.remainder.box.width}`)
    .attr("height", `${guidesModel.remainder.box.height}`)
    .attr("patternUnits", "userSpaceOnUse")
    .attr("patternTransform", "rotate(20)");

  extraPattern
    .append("text")
    .text(`${dataModel.currentQuarter.wholeWeeksLeft.durationInWeeks}`)
    .attr("x", 0)
    .attr("y", `${dpr(guidesModel.remainder.box.textY)}`)
    .attr("style", `font-size: ${guidesModel.remainder.fontSize}`)
    .attr("opacity", "0.9")
    .attr("fill", guidesModel.colors.remainder.weekNumber.color);

  const arcGenerator = defaultArcGenerator();

  arcGenerator
    .innerRadius(guidesModel.outerRadius / 3)
    .outerRadius(guidesModel.outerRadius - guidesModel.outerRadius / 15);

  const remainderArc = {
    startAngle:
      dataModel.currentQuarter.wholeWeeksLeft.start.yearFraction *
      2.0 *
      Math.PI,
    endAngle:
      dataModel.currentQuarter.wholeWeeksLeft.end.yearFraction * 2.0 * Math.PI,
  };

  const remainderArcs = dataModel.currentQuarter.availableDays.map((u) => {
    return {
      startAngle: u.start.yearFraction * 2.0 * Math.PI,
      endAngle: u.end.yearFraction * 2.0 * Math.PI,
    };
  });

  const parentGroup = svg.selection.append("g").attr("id", "remainder");

  parentGroup
    .append("clipPath")
    .attr("id", "remainder-clip")
    .selectAll()
    .data(remainderArcs)
    .enter()
    .append("path")
    .attr("d", (arc) => arcGenerator(arc))
    .attr("stroke", "none");

  parentGroup
    .append("path")
    .attr("class", "remainder")
    .attr("d", arcGenerator(remainderArc))
    .attr("fill", "url(#remainder-pattern)")
    .attr("clip-path", "url(#remainder-clip)")
    .attr("stroke", "none");

  parentGroup
    .append("path")
    .attr("class", "remainder")
    .attr("d", arcGenerator(remainderArc))
    .attr("fill", "url(#remainder-pattern-extra)")
    .attr("clip-path", "url(#remainder-clip)")
    .attr("stroke", "none");
}

export function drawRemainderLabel(dataModel, guidesModel, svg) {
  const parentGroup = svg.selection.append("g").attr("id", "remainder-label");

  const positionForQuarterIndex = [
    guidesModel.remainderLabel.left,
    guidesModel.remainderLabel.left,
    guidesModel.remainderLabel.right,
    guidesModel.remainderLabel.right,
  ];

  const position = positionForQuarterIndex[dataModel.currentQuarter.index];

  const remaining = dataModel.currentQuarter.wholeWeeksLeft.durationInWeeks;
  parentGroup
    .append("text")
    .text(`${remaining} week${remaining === 1 ? "" : "s"} left`)
    .attr("x", dpr(position.x))
    .attr("y", dpr(position.y))
    .attr(
      "style",
      `font-size: ${guidesModel.remainderLabel.fontSize}; dominant-baseline: hanging; text-anchor: left`
    )
    .attr("fill", guidesModel.colors.remainder.color)
    .attr("stroke", "none");
}
