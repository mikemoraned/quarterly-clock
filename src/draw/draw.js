import {
  drawCurrentQuarter,
  drawAllQuarterContext,
  drawQuarterLabel,
} from "./quarter";
import { drawWeekScale, drawDayHand } from "./year";
import { drawLogo, drawReadme } from "./about";
import { drawRemainder, drawRemainderLabel } from "./remainder";
import { defaultFontSizeFormat, defaultPositionRounding as dpr } from "./standard";

// color blind safe colors from Bang Wong in and https://www.nature.com/articles/nmeth.1618
// and https://davidmathlogic.com/colorblind
const BANG_WONG_PALETTE = {
  black: "rgb(0, 0, 0)",
  orange: "rgb(230, 159, 0)",
  "sky-blue": "rgb(86, 180, 233)",
  "bluish-green": "rgb(0, 158, 115)",
  yellow: "rgb(240, 228, 66)",
  blue: "rgb(0, 114, 178)",
  vermillion: "rgb(213, 94, 0)",
  "reddish-purple": "rgb(204, 121, 167)",
};

export function createGuidesModel(svg) {
  const outerMargin = 100;

  const sideLength = Math.min(
    svg.width - outerMargin,
    svg.height - outerMargin
  );
  const clockRadius = sideLength / 2.0;

  const guidesModel = {
    outerRadius: clockRadius,
    remainder: {
      fontSize: defaultFontSizeFormat(sideLength / 15),
      box: {
        width: 45 * ((0.5 * sideLength) / 275),
        height: 35 * ((0.5 * sideLength) / 275),
        textY: 30 * ((0.5 * sideLength) / 275),
      },
    },
    colors: {
      quarters: [
        {
          color: BANG_WONG_PALETTE["yellow"],
        },
        {
          color: BANG_WONG_PALETTE["reddish-purple"],
        },
        {
          color: BANG_WONG_PALETTE["sky-blue"],
        },
        {
          color: BANG_WONG_PALETTE["bluish-green"],
        },
      ],
      remainder: {
        bg: {
          color: BANG_WONG_PALETTE["vermillion"],
        },
        weekNumber: {
          color: "black",
        },
      },
    },
    quarterLabel: {
      fontSize: defaultFontSizeFormat(sideLength / 5),
      left: {
        x: -1.0 * clockRadius * 0.9,
        y: 0,
      },
      right: {
        x: clockRadius * 0.35,
        y: 0,
      },
    },
    remainderLabel: {
      fontSize: defaultFontSizeFormat(sideLength / 22),
      left: {
        x: -1.0 * clockRadius * 0.87,
        y: clockRadius * 0.03,
      },
      right: {
        x: clockRadius * 0.38,
        y: clockRadius * 0.03,
      },
    },
    logo: {
      fontSize: defaultFontSizeFormat(sideLength / 13),
      top: {
        x: 0,
        y: -1.0 * clockRadius * 0.6,
      },
      bottom: {
        x: 0,
        y: clockRadius * 0.4,
      },
    },
    readme: {
      fontSize: defaultFontSizeFormat(sideLength / 30),
      top: {
        x: 0,
        y: -1.0 * clockRadius * 0.47,
      },
      bottom: {
        x: 0,
        y: clockRadius * 0.52,
      },
    },
    weekScale: {
      fontSize: sideLength / 45,
      yOffset: ((6 / 14) * sideLength) / 45,
    },
  };

  return guidesModel;
}

export function draw(dataModel, svg) {
  const guidesModel = createGuidesModel(svg);

  console.log("Guides Model:", JSON.stringify(guidesModel));

  drawWeekScale(dataModel, guidesModel, svg);
  drawAllQuarterContext(dataModel, guidesModel, svg);
  drawCurrentQuarter(dataModel, guidesModel, svg);
  if (dataModel.currentQuarter.wholeWeeksLeft.durationInWeeks != 0) {
    drawRemainder(dataModel, guidesModel, svg);
  }
  drawRemainderLabel(dataModel, guidesModel, svg);
  drawDayHand(dataModel, guidesModel, svg);
  drawQuarterLabel(dataModel, guidesModel, svg);
  drawLogo(dataModel, guidesModel, svg);
  drawReadme(dataModel, guidesModel, svg);
  drawGuides(guidesModel, svg);
}

function drawGuides(model, svg) {
  const parentGroup = svg.selection.append("g").attr("id", "guides");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", 0)
    .attr("cy", 0)
    .attr("r", dpr(model.outerRadius))
    .style("fill", "none")
    .style("stroke", "red");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", dpr(model.quarterLabel.left.x))
    .attr("cy", dpr(model.quarterLabel.left.y))
    .attr("r", 5)
    .style("fill", "red")
    .style("stroke", "black");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", dpr(model.quarterLabel.right.x))
    .attr("cy", dpr(model.quarterLabel.right.y))
    .attr("r", 5)
    .style("fill", "red")
    .style("stroke", "black");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", dpr(model.remainderLabel.left.x))
    .attr("cy", dpr(model.remainderLabel.left.y))
    .attr("r", 5)
    .style("fill", "red")
    .style("stroke", "black");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", dpr(model.remainderLabel.right.x))
    .attr("cy", dpr(model.remainderLabel.right.y))
    .attr("r", 5)
    .style("fill", "red")
    .style("stroke", "black");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", dpr(model.logo.top.x))
    .attr("cy", dpr(model.logo.top.y))
    .attr("r", 5)
    .style("fill", "red")
    .style("stroke", "black");

  parentGroup
    .append("circle")
    .attr("class", "guide")
    .attr("cx", dpr(model.logo.bottom.x))
    .attr("cy", dpr(model.logo.bottom.y))
    .attr("r", 5)
    .style("fill", "red")
    .style("stroke", "black");
}
