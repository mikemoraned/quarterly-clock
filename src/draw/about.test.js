import { modelForDate } from "../model.js";
import { createGuidesModel } from "./draw.js";
import { drawLogo, drawReadme } from "./about.js";
import { addSvgRoot } from "./root.js";
import { createParentNode, takeSnapshot, DIMENSIONS } from "./test_setup.js";
import { describe, test, expect } from 'vitest';

describe("drawLogo", () => {
  function render(model) {
    const parentNode = createParentNode();
    const svg = addSvgRoot(parentNode, DIMENSIONS);
    const guidesModel = createGuidesModel(svg);

    drawLogo(model, guidesModel, svg);

    return parentNode;
  }

  test("start of Q1", () => {
    const now = Date.parse("01 Jan 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q1", () => {
    const now = Date.parse("31 Mar 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("start of Q2", () => {
    const now = Date.parse("01 Apr 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q2", () => {
    const now = Date.parse("30 Jun 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("start of Q3", () => {
    const now = Date.parse("01 Jul 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q3", () => {
    const now = Date.parse("30 Sep 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("start of Q4", () => {
    const now = Date.parse("01 Oct 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q4", () => {
    const now = Date.parse("31 Dec 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });
});

describe("drawReadme", () => {
  function render(model) {
    const parentNode = createParentNode();
    const svg = addSvgRoot(parentNode, DIMENSIONS);
    const guidesModel = createGuidesModel(svg);

    drawReadme(model, guidesModel, svg);

    return parentNode;
  }

  test("start of Q1", () => {
    const now = Date.parse("01 Jan 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q1", () => {
    const now = Date.parse("31 Mar 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("start of Q2", () => {
    const now = Date.parse("01 Apr 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q2", () => {
    const now = Date.parse("30 Jun 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("start of Q3", () => {
    const now = Date.parse("01 Jul 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q3", () => {
    const now = Date.parse("30 Sep 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("start of Q4", () => {
    const now = Date.parse("01 Oct 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });

  test("last day of Q4", () => {
    const now = Date.parse("31 Dec 2021 00:00:00 GMT");
    const model = modelForDate(now);

    const parentNode = render(model);

    expect(takeSnapshot(parentNode)).toMatchSnapshot();
  });
});
