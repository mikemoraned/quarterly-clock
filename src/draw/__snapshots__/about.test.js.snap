// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`drawLogo last day of Q1 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,270 A 450,450 0 0,1 225,270" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo last day of Q2 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,-180 A 450,450 0 0,1 225,-180" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo last day of Q3 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,-180 A 450,450 0 0,1 225,-180" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo last day of Q4 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,270 A 450,450 0 0,1 225,270" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo start of Q1 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,270 A 450,450 0 0,1 225,270" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo start of Q2 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,-180 A 450,450 0 0,1 225,-180" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo start of Q3 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,-180 A 450,450 0 0,1 225,-180" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawLogo start of Q4 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="logo">
            <defs>
                <path d="M -225,270 A 450,450 0 0,1 225,270" id="logoPath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock">
                <text style="font-size: 69.2307692307px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#logoPath">
                        <tspan style="font-family: &quot;Font Awesome 5 Free&quot;">
                            
                        </tspan>
                        <tspan>
                            uarterly
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme last day of Q1 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,324 A 450,450 0 0,1 225,324" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme last day of Q2 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,-121.5 A 450,450 0 0,1 225,-121.5" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme last day of Q3 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,-121.5 A 450,450 0 0,1 225,-121.5" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme last day of Q4 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,324 A 450,450 0 0,1 225,324" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme start of Q1 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,324 A 450,450 0 0,1 225,324" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme start of Q2 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,-121.5 A 450,450 0 0,1 225,-121.5" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme start of Q3 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,-121.5 A 450,450 0 0,1 225,-121.5" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;

exports[`drawReadme start of Q4 1`] = `
"<svg id="clock" width="1000" height="1000">
    <g transform="translate(500 500)">
        <g id="readme">
            <defs>
                <path d="M -225,324 A 450,450 0 0,1 225,324" id="readmePath"></path>
            </defs>
            <a href="https://github.com/mikemoraned/quarterly-clock/blob/main/README.md">
                <text style="font-size: 30px; dominant-baseline: middle; text-anchor: middle" fill="black">
                    <textPath startOffset="50%" href="#readmePath">
                        <tspan style="font-family: &quot;Font Awesome 5 Brands&quot;">
                            
                        </tspan>
                        <tspan>
                            Readme
                        </tspan>
                    </textPath>
                </text>
            </a>
        </g>
    </g>
</svg>"
`;
